csharp/src/Apache.Arrow/Builder/BufferBuilder.cs

```diff
@@ -0,0 +1,356 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Apache.Arrow.Memory;
+
+namespace Apache.Arrow.Builder
+{
+    public class BufferBuilder : IBufferBuilder
+    {
+        public class BitBuffer
+        {
+            private readonly bool[] _bits;
+
+            public int Length { get; private set; }
+            public int AvailableLength => Capacity - Length;
+
+            public int Capacity;
+
+            public bool IsFull => Length == Capacity;
+            public byte ToByte(ref byte data) => BitUtility.ToByte(ref data, _bits);
+
+            public BitBuffer(int capacity = 8)
+            {
+                Capacity = capacity;
+                _bits = new bool[capacity];
+                Length = 0;
+            }
+
+            public void Append(bool bit) => _bits[Length++] = bit;
+            public void Fill(ReadOnlySpan<bool> bits)
+            {
+                bits.CopyTo(_bits.AsSpan().Slice(Length, bits.Length));
+                Length += bits.Length;
+            }
+
+            public void Reset()
+            {
+                for (int i = 0; i < _bits.Length; i++)
+                {
+                    _bits[i] = false;
+                }
+                Length = 0;
+            }
+        }
+
+        private const int DefaultCapacity = 64;
+        public int ByteLength { get; private set; }
+
+        public Memory<byte> Memory { get; private set; }
+        public BitBuffer BitOverhead { get; }
```

Comment:
This approach of using a BitBuffer for BitOverhead is inefficient because it is constantly allocating small byte spans.  I created a quick benchmark comparing this with a simpler approach for demonstration though its possible I'm missing something or did something incorrect:

https://gist.github.com/westonpace/35b619c4c9e90824d93ecb8306a9e169

```
|                Method | NumBits |          Mean |        Error |     StdDev |
|---------------------- |-------- |--------------:|-------------:|-----------:|
|           AppendFalse |       1 |      17.82 ns |     0.101 ns |   0.095 ns |
|            AppendTrue |       1 |      19.65 ns |     0.317 ns |   0.296 ns |
| AppendFalseNoOverhead |       1 |      10.42 ns |     0.072 ns |   0.067 ns |
|  AppendTrueNoOverhead |       1 |      10.30 ns |     0.100 ns |   0.089 ns |
|           AppendFalse |     100 |     257.41 ns |     1.035 ns |   0.968 ns |
|            AppendTrue |     100 |     293.54 ns |     1.179 ns |   1.102 ns |
| AppendFalseNoOverhead |     100 |      36.18 ns |     0.284 ns |   0.266 ns |
|  AppendTrueNoOverhead |     100 |      34.80 ns |     0.094 ns |   0.088 ns |
|           AppendFalse |   10000 |  22,638.86 ns |    77.641 ns |  68.827 ns |
|            AppendTrue |   10000 |  25,443.68 ns |    65.197 ns |  60.985 ns |
| AppendFalseNoOverhead |   10000 |      49.20 ns |     0.159 ns |   0.141 ns |
|  AppendTrueNoOverhead |   10000 |      49.24 ns |     0.180 ns |   0.168 ns |
|           AppendFalse |  100000 | 235,708.02 ns |   633.895 ns | 592.946 ns |
|            AppendTrue |  100000 | 271,133.50 ns | 1,123.712 ns | 996.141 ns |
| AppendFalseNoOverhead |  100000 |     346.36 ns |     0.456 ns |   0.380 ns |
|  AppendTrueNoOverhead |  100000 |     351.31 ns |     1.170 ns |   1.094 ns |
```

Reply:
During its lifetime, the builder will allocate only 1 time 8 bits in bit overhead

On Ever bit(s) append, it will check if the 8 bits are written (like the old bitmap builder checking if it is % 8)

If it is full (8 bits are written) i will write a new byte in the Memory Buffer and reset all values in the BitOverhead to false setting his length to 0 an dread to recieve other bits

```csharp
public class BenchmarkBits
{
    public static long ElapsedTicks(int repetition, Action action)
    {
        Stopwatch stopwatch = Stopwatch.StartNew();

        for (int i = 0; i < repetition; i++)
            action();

        stopwatch.Stop();

        return stopwatch.ElapsedTicks / repetition;
    }

    private readonly ITestOutputHelper output;

    private static readonly bool[] trueBits = Enumerable.Range(0, 100).Select(_ => true).ToArray();
    private static readonly bool[] falseBits = Enumerable.Range(0, 100).Select(_ => false).ToArray();

    public BenchmarkBits(ITestOutputHelper output)
    {
        this.output = output;
    }

    [Fact]
    public void Bench()
    {
        output.WriteLine($"Elapsed {ElapsedTicks(1000000, MakeOld)}");
        output.WriteLine($"Elapsed {ElapsedTicks(1000000, MakeNew)}");
    }

    private static void MakeNew()
    {
        var builder = new BufferBuilder(64);

        builder.AppendBit(true).AppendBit(true)
            .AppendBits(trueBits)
            .AppendBits(falseBits)
            .Build();
    }

    private static void MakeOld()
    {
        var builder = new ArrowBuffer.BitmapBuilder(64);

        builder.Append(true).Append(true)
            .AppendRange(trueBits)
            .AppendRange(falseBits)
            .Build();
    }
}
````

On 1 million iterations the new implementation is in mean 168ticks
On 1 million iterations the old implementation is in mean 312 ticks

Reply:
Same for sequential append it looks better, thus it scales good with larger value range

Old was in mean 1183 Ticks vs 938 Ticks for the new

```csharp
[Fact]
public void Bench()
{
    output.WriteLine($"Elapsed {ElapsedTicks(100000, MakeOld)}");
    output.WriteLine($"Elapsed {ElapsedTicks(100000, MakeNew)}");
}

private static void MakeNew()
{
    var builder = new BufferBuilder(64);

    for (int i = 0; i < 1000; i++)
        builder.AppendBit(true);

    builder.Build();
}

private static void MakeOld()
{
    var builder = new ArrowBuffer.BitmapBuilder(64);

    for (int i = 0; i < 1000; i++)
        builder.Append(true);

    builder.Build();
}
```